analyses/malawi/scripts/mwi_chirps_dry_spell_detection.R

```diff
@@ -0,0 +1,127 @@
+s2011 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2011_p25.nc")
+s2012 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2012_p25.nc")
+s2013 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2013_p25.nc")
+s2014 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2014_p25.nc")
+s2015 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2015_p25.nc")
+s2016 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2016_p25.nc")
+s2017 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2017_p25.nc")
+s2018 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2018_p25.nc")
+s2019 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2019_p25.nc")
+s2020 <- raster::stack("../../indicators/drought/Data/chirps/chirps_global_daily_2020_p25.nc")
+
+s2010_s2020 <- stack(s2010, s2011, s2012, s2013, s2014, s2015, s2016, s2017, s2018, s2019, s2020) # all files combined into a stack
+
+# crop and masked area outside of MWI
+s2010_s2020_cropped <- crop(x = s2010_s2020, y = extent(mwi_adm2_spatial_extent)) # crop converts to a brick - a single raster file
+data <- mask(s2010_s2020_cropped, mask = mwi_adm2)
+
+plot(data) # visual inspection
+
+# explore compiled raster file ("brick")
+st_crs(data) # coordinate reference system
+st_bbox(data) # spatial extent
+ncell(data) # number of cells per layer (nrow * ncol)
+nrow(data) # number of rows in a layer
+ncol(data) # number of columns in a layer
+nlayers(data) # number of layers (days)
+dim(data) # (nrow, ncol, nlayers)
+yres(data) # y-resolution
+xres(data) # x-resolution
+
+data_projection <- projection(data) # assign CRS to variable
+
+# create list of regions
+region_list <- mwi_adm2[,c('ADM2_PCODE', 'ADM2_EN', 'geometry')]
+
+## compute precipitation totals per adm2
+
+# loop through layers/days to compile MAX values across layers/days
+nbr_layers <- nlayers(data)
+data_max_values <- data.frame(ID = 1:nrow(mwi_adm2))
+
+for (i in seq_along(1:nbr_layers)) {
+  
+        data_max_values <- computeLayerStat(i, max, data_max_values)
+        
+      }
+
+# compute per-region 14-d rolling sums
+data_max_sums <- compute14dSum(data_max_values)
+
+## identify dry spells per adm2
+
+# list days on which 14-day rolling sum is 2mm or less of rain
+data_max_sums$rollsum_ds_bin <- ifelse(data_max_sums$rollsum_14d <= 2, 1, 0)
+
+# identify beginning, end and duration of dry spells per region
+dry_spells_list <- data_max_sums %>%
+                      mutate(rollsum_ds_bin = ifelse(is.na(rollsum_ds_bin), 0, rollsum_ds_bin)) %>%  # remove NAs with 0
+                      group_by(pcode, spell = cumsum(c(0, diff(rollsum_ds_bin) != 0))) %>% # groups consecutive days with rolling sum <= 2mm
```

Comment:
Regarding the minimum reported dry spell being 15 days while you would expect there to be also 14 day dryspells: could it be that this line should be `cumsum(c(0, diff(rollsum_ds_bin) **== 1**)))` ? I am not fully understanding the code yet, but in `rollsum_ds_bin` there are also -1's which are now counted in the same way as 1. If I change this line, I do get 14 day period dryspells. And an example where a 14 day period dryspell occurred is  on `last_dry_spell_date=2010-06-30` in MW101

Reply:
Hm. I only have 1's and 0's under `rollsum_ds_bin`. That variable flag days on which the 14-day trailing sum is less then or equal to 2mm (ie a binary indicating if the rolling sum indicates a dry spell).

Ln104 is admittedly hairy. It's essentially assigning a "streak number" to series of rows that share a `rollsum_ds_bin` value. The series for which `rollsum_ds_bin` == 1 are dry spells since the rolling sum is over 14 days. 

I think I found the bug: Ln105. It should read n() >= 1. In the case you pointed out, there was a single date on which the 14-day sum was <= 2mm (ie `rollsum_ds_bin == 1`). The dry spell was exactly 14 days. Because it's only on its last day that the definition of dry spell was met, it was a single record of a dry spell and that case was filtered out by n() > 1.

Reply:
Let me know if this works for you (no -1's under `rollsum_ds_bin`, presence of 14-day spells, and flagging of the streak ending on 2010-06-30 in MW101 as a dry spell. 

Thanks for helping figure this out!

Reply:
Oops yes, something went wrong with the minusses in me poking the code, but think the n>=1 makes sense